The study I am conducting will be analyzing the Instacart orders placed in 2017. The purpose is to clean up the raw data and give insight into the shopping habits of Instacart customers. First, I will open all of the csv files starting with instacart_orders.csv:

In [1]:
#Open and read instacart_orders.csv
import pandas as pd
import numpy as np


In [2]:
orders_df = pd.read_csv('/datasets/instacart_orders.csv',sep=';')
orders_df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478967 entries, 0 to 478966
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   order_id                478967 non-null  int64  
 1   user_id                 478967 non-null  int64  
 2   order_number            478967 non-null  int64  
 3   order_dow               478967 non-null  int64  
 4   order_hour_of_day       478967 non-null  int64  
 5   days_since_prior_order  450148 non-null  float64
dtypes: float64(1), int64(5)
memory usage: 21.9 MB


In [3]:
display(orders_df.head)

<bound method NDFrame.head of         order_id  user_id  order_number  order_dow  order_hour_of_day  \
0        1515936   183418            11          6                 13   
1        1690866   163593             5          5                 12   
2        1454967    39980             4          5                 19   
3        1768857    82516            56          0                 20   
4        3007858   196724             2          4                 12   
...          ...      ...           ...        ...                ...   
478962   3210681     5617             5          1                 14   
478963   3270802   112087             2          3                 13   
478964    885349    82944            16          2                 11   
478965    216274     4391             3          3                  8   
478966   2071924     1730            18          1                 14   

        days_since_prior_order  
0                         30.0  
1                          

In [4]:
#Open and read products.csv
products_df = pd.read_csv('/datasets/products.csv', sep=';')
products_df.info(show_counts=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49694 entries, 0 to 49693
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49694 non-null  int64 
 1   product_name   48436 non-null  object
 2   aisle_id       49694 non-null  int64 
 3   department_id  49694 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


In [5]:
display(products_df.head)

<bound method NDFrame.head of        product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49689       49690                      HIGH PERFORMANCE ENERGY DRINK   
49690       49691                      ORIGINAL PANCAKE & WAFFLE MIX   
49691       49692    ORGANIC INSTANT OATMEAL LIGHT MAPLE BROWN SUGAR   
49692       49693                             SPRING WATER BODY WASH   
49693       49694                            BURRITO- STEAK & CHEESE   

       aisle_id  department_id  
0            61             19  
1           104             13  
2     

In [6]:
#Open and read order_products.csv
order_products_df = pd.read_csv('/datasets/order_products.csv', sep=';')
order_products_df.info(show_counts=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4545007 entries, 0 to 4545006
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   order_id           4545007 non-null  int64  
 1   product_id         4545007 non-null  int64  
 2   add_to_cart_order  4544171 non-null  float64
 3   reordered          4545007 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 138.7 MB


In [7]:
display(order_products_df.head)

<bound method NDFrame.head of          order_id  product_id  add_to_cart_order  reordered
0         2141543       11440               17.0          0
1          567889        1560                1.0          1
2         2261212       26683                1.0          1
3          491251        8670               35.0          1
4         2571142        1940                5.0          1
...           ...         ...                ...        ...
4545002    577211       15290               12.0          1
4545003   1219554       21914                9.0          0
4545004    692640       47766                4.0          1
4545005    319435         691                8.0          1
4545006   1398151       28733                9.0          0

[4545007 rows x 4 columns]>

In [8]:
#Open and read aisles.csv
aisles_df = pd.read_csv('/datasets/aisles.csv', sep=';')
aisles_df.info(show_counts=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   aisle_id  134 non-null    int64 
 1   aisle     134 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.2+ KB


In [9]:
display(aisles_df.head)

<bound method NDFrame.head of      aisle_id                       aisle
0           1       prepared soups salads
1           2           specialty cheeses
2           3         energy granola bars
3           4               instant foods
4           5  marinades meat preparation
..        ...                         ...
129       130    hot cereal pancake mixes
130       131                   dry pasta
131       132                      beauty
132       133  muscles joints pain relief
133       134  specialty wines champagnes

[134 rows x 2 columns]>

In [10]:
#Open and read departments.csv
departments_df = pd.read_csv('/datasets/departments.csv', sep=';')
departments_df.info(show_counts=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  21 non-null     int64 
 1   department     21 non-null     object
dtypes: int64(1), object(1)
memory usage: 464.0+ bytes


In [11]:
display(departments_df.head)

<bound method NDFrame.head of     department_id       department
0               1           frozen
1               2            other
2               3           bakery
3               4          produce
4               5          alcohol
5               6    international
6               7        beverages
7               8             pets
8               9  dry goods pasta
9              10             bulk
10             11    personal care
11             12     meat seafood
12             13           pantry
13             14        breakfast
14             15     canned goods
15             16       dairy eggs
16             17        household
17             18           babies
18             19           snacks
19             20             deli
20             21          missing>

The instacart_orders.csv is broken down to unique information for each order made.
The products.csv is broken down into unique products that customers can buy.
The order_products.csv is broken down into every item that customers can buy.
The aisles.csv breaks down every grocery item by aisle and its own unique identifier.
The departments.csv groups all the groceries into departments and gives them unique identifiers for each department.

Overall there appears to be 478,967 orders of the 49,694 products within the 134 aisles and 21 departments. 

In [12]:
#use .fillna to replace NA data in ID column into numbers in order to convert column to integers
orders_df['order_id'].fillna(-1, inplace=True)
orders_df['user_id'].fillna(-1, inplace=True)

products_df['product_id'].fillna(-1, inplace=True)

aisles_df['aisle_id'].fillna(-1, inplace=True)

departments_df['department_id'].fillna(-1, inplace=True)

order_products_df['add_to_cart_order'].fillna(-1, inplace=True)
order_products_df['add_to_cart_order'] = order_products_df['add_to_cart_order'].replace([np.inf, -np.inf], -1)



In [13]:
#convert ID columns to integers

orders_df['order_id'] = orders_df['order_id'].astype(int)
orders_df['user_id'] = orders_df['user_id'].astype(int)

products_df['product_id'] = products_df['product_id'].astype(int)

aisles_df['aisle_id'] = aisles_df['aisle_id'].astype(int)

departments_df['department_id'] = departments_df['department_id'].astype(int)

order_products_df['order_id'] = order_products_df['order_id'].astype(int)
order_products_df['product_id'] = order_products_df['product_id'].astype(int)

#check conversion to integer worked
print(orders_df['order_id'].dtype)
print(orders_df['user_id'].dtype)

print(order_products_df['order_id'].dtype)
print(order_products_df['product_id'].dtype)

int64
int64
int64
int64


In [14]:
#finding the missing values

print(orders_df.isna().sum())
print()
print(products_df.isna().sum())
print()
print(order_products_df.isna().sum())
print()
print(aisles_df.isna().sum())
print()
print(departments_df.isna().sum())



order_id                      0
user_id                       0
order_number                  0
order_dow                     0
order_hour_of_day             0
days_since_prior_order    28819
dtype: int64

product_id          0
product_name     1258
aisle_id            0
department_id       0
dtype: int64

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

aisle_id    0
aisle       0
dtype: int64

department_id    0
department       0
dtype: int64


In [15]:
#handling the missing values
orders_df=orders_df.dropna()
print(orders_df.isna().sum())
print()
products_df=products_df.dropna()
print(products_df.isna().sum())

order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

product_id       0
product_name     0
aisle_id         0
department_id    0
dtype: int64


## Find and remove duplicate values (and describe why you make your choices)

### `orders` data frame

In [16]:
# Check for how many/if any duplicated orders are in the data sets

duplicates = orders_df.duplicated().sum()
print(f"Number of duplicate rows in orders: {duplicates}")

duplicates = products_df.duplicated().sum()
print(f"Number of duplicate rows in products: {duplicates}")

duplicates = order_products_df.duplicated().sum()
print(f"Number of duplicate rows in order_products: {duplicates}")

duplicates = aisles_df.duplicated().sum()
print(f"Number of duplicate rows in aisles: {duplicates}")

duplicates = departments_df.duplicated().sum()
print(f"Number of duplicate rows in departments: {duplicates}")

Number of duplicate rows in orders: 13
Number of duplicate rows in products: 0
Number of duplicate rows in order_products: 0
Number of duplicate rows in aisles: 0
Number of duplicate rows in departments: 0


In [17]:
#now checking the order_df data set for what the 13 duplicates are
duplicates = orders_df[orders_df.duplicated(keep=False)]
print(duplicates)

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
30371    1918001   188546            14          3                  2   
99462     794638    50898            24          3                  2   
119251    408114    68324             4          3                  2   
142258   2845099    31189            11          3                  2   
145574    794638    50898            24          3                  2   
164581   1112182   202304            84          3                  2   
204042   2125197    14050            48          3                  2   
215294   2160484   107525            16          3                  2   
223105   2160484   107525            16          3                  2   
230807   1918001   188546            14          3                  2   
242618   1021560    53767             3          3                  2   
250626    467134    63189            21          3                  2   
257934   1919531   191501            32          3 

I found duplicates only in the instacart_orders.csv file. After reviewing, I found the only duplicates are the day of the week and time of day that the orders were placed. All the other information in the other columns are different incliding the custoemr id numbers which leads me to conclude these are not duplicates and are in fact seperate orders. With this information, I am leaving the 'duplicates' inside the dataframe.

In [18]:
#Finding which day is Wednesday
print(orders_df['order_dow'].unique())

min_dow = orders_df['order_dow'].min()
max_dow = orders_df['order_dow'].max()

print(f"Minimum day of week: {min_dow}")
print(f"Maximum day of week: {max_dow}")


print()
#checking the time that is 2am
print(orders_df['order_hour_of_day'].unique())

min_hour = orders_df['order_hour_of_day'].min()
max_hour = orders_df['order_hour_of_day'].max()

print(f"Minimum hour of day: {min_hour}")
print(f"Maximum hour of day: {max_hour}")


[6 5 0 4 3 1 2]
Minimum day of week: 0
Maximum day of week: 6

[13 12 19 20 11  8 18 15 16  9 22 10  0 14 17  1  7  6  2 21  3 23  5  4]
Minimum hour of day: 0
Maximum hour of day: 23


For this I am assuming Wednesday is 2 based on an education guess. A google search states that Sundays are the busiest days of the week for orders and their pay period is from Monday-Sunday. This would make Sunday 6, which is the busiest day in my data, and Wednesday would be 2. See https://www.bing.com/search?q=what+is+instacarts+pay+period+per+week and https://www.bing.com/search?q=what+are+instacarts+busiest+day for more information.

In [19]:
# Check for all orders placed Wednesday at 2:00 AM
wednesday_orders = orders_df.loc[(orders_df['order_dow'] == 3) & (orders_df['order_hour_of_day'] == 15)]
print(wednesday_orders)

        order_id  user_id  order_number  order_dow  order_hour_of_day  \
16       3109465    55090            19          3                 15   
30       1890318   193501             3          3                 15   
66       3335598   183111            86          3                 15   
178      2138117    52728             9          3                 15   
212      2946607    17731             5          3                 15   
...          ...      ...           ...        ...                ...   
478546    342498    70253            13          3                 15   
478570   1712187    57261             4          3                 15   
478598    726771   200005            20          3                 15   
478766     83789   162505            19          3                 15   
478898   2696462    15409             8          3                 15   

        days_since_prior_order  
16                         9.0  
30                        30.0  
66                      

In [20]:
# Remove duplicate orders
orders_df.drop_duplicates(inplace=True)

In [21]:
# Double check for duplicate rows
duplicate_rows = orders_df[orders_df.duplicated()]
print(duplicate_rows)
duplicate_counts = orders_df.duplicated().value_counts()
print(duplicate_counts)

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []
False    450135
dtype: int64


I have looked for duplicate orders, removed them, and then checked to make sure I successfully removed them

In [22]:
# Double check for duplicate order IDs only
duplicate_order_ids = orders_df[orders_df.duplicated(subset='order_id', keep=False)]
print(duplicate_order_ids)
print()
duplicate_order_id_counts = orders_df['order_id'].value_counts()
print(duplicate_order_id_counts[duplicate_order_id_counts > 1].index)

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Int64Index([], dtype='int64')


After revewing the duplicate orders for IDs only, I have conlcluded there are no duplicate orders for ID's. This leads me to conlclude there are no duplicate orders in this data, assuming that all of the orders have their own unique ID's.

### `products` data frame

In [23]:
# Check for fully duplicate rows
duplicates = products_df.duplicated(keep=False)
print(products_df[duplicates])
print()
print(products_df)
print()
print(products_df.duplicated())
print()
print(products_df.duplicated().sum())


Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id]
Index: []

       product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49689       49690                      HIGH PERFORMANCE ENERGY DRINK   
49690       49691                      ORIGINAL PANCAKE & WAFFLE MIX   
49691       49692    ORGANIC INSTANT OATMEAL LIGHT MAPLE BROWN SUGAR   
49692       49693                             SPRING WATER BODY WASH   
49693       49694                            BURRITO- STEAK & CHEESE   

       aisle_id  department_id  
0            6

This shows there are no fully duplicates rows

In [24]:
# Check for just duplicate product IDs
duplicate_product_ids = products_df[products_df.duplicated(subset='product_id', keep=False)]
print(duplicate_product_ids)
print()
print()
print(products_df.duplicated())
print()
print(products_df.duplicated().sum())


Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id]
Index: []


0        False
1        False
2        False
3        False
4        False
         ...  
49689    False
49690    False
49691    False
49692    False
49693    False
Length: 48436, dtype: bool

0


This concludes there are no duplicates in the product_ids columns of the product.df dataframe

In [ ]:
# Check for just duplicate product names (convert names to lowercase to compare better)
if 'product_name' in products_df.columns:
    duplicate_product_names = products_df[products_df['product_name'].str.lower().duplicated(keep=False)]
    print(duplicate_product_names)
else:
    print("The column 'product_name' does not exist in the DataFrame.")



This suggests that there are 207 duplicate product names in the products_df dataframe.

In [ ]:
# Check for duplicate product names that aren't missing

if 'product_name' in products_df.columns:
    duplicate_product_names = products_df[(products_df['product_name'].notna()) & 
                                        (products_df['product_name'].str.lower().duplicated(keep=False))]
    print(duplicate_product_names)
else:
    print("The column 'product_name' does not exist in the DataFrame.")
    


This suggests that there are 207 duplicate product names in the products_df dataframe.

### `departments` data frame

In [ ]:
#I am guessing this is asking for me to check the departments_df dataframe for department_id duplicates

print(departments_df)
print()
duplicates = departments_df[departments_df['department_id'] & departments_df['department_id'].duplicated(keep=False)]
print(duplicates)


In [ ]:
#I am guessing this is asking for me to check the departments_df dataframe for department_id duplicates, not including the missing values
duplicates = departments_df[departments_df['department_id'].notna() & departments_df['department_id'].duplicated(keep=False)]
print(duplicates)



This concludes there are no duplicates of the department_ids in the data frame.

### `aisles` data frame

In [ ]:
#I am guessing this is asking for me to check the aisles_df dataframe for aisles_id duplicates
print(aisles_df)
duplicates = aisles_df[aisles_df['aisle_id'] & aisles_df['aisle_id'].duplicated(keep=False)]
print(duplicates)

In [ ]:
duplicates = aisles_df[aisles_df['aisle_id'].notna() & aisles_df['aisle_id'].duplicated(keep=False)]
print(duplicates)



This concludes that there are no duplicates of the aisle_id in the dataframe.

### `order_products` data frame

In [ ]:
# Check for fullly duplicate rows
duplicates = order_products_df[order_products_df.duplicated(keep=False)]
print(duplicates)

In [ ]:
# Double check for any other tricky duplicates
tricky_duplicates = order_products_df[order_products_df.duplicated(subset=['product_id', 'add_to_cart_order', 'reordered'], keep=False)]
print(tricky_duplicates)
print(duplicates.shape[0])
print(tricky_duplicates.shape[0])

This shows there are no duplicates or tricky duplicates in the dataframe order_products 

## Find and remove missing values


### `products` data frame

In [ ]:
# Are all of the missing product names associated with aisle ID 100?
if 'product_name' in products_df.columns:
    duplicate_product_names = products_df[products_df['product_name'].str.lower().duplicated(keep=False)]
    aisle_ids_of_duplicates = duplicate_product_names['aisle_id'].unique()
    print(aisle_ids_of_duplicates)
print()
print("This suggests that the missing product names are not exclusively associated with aisle ID 100.")

In [ ]:
# Are all of the missing product names associated with department ID 21?
if 'product_name' in products_df.columns:
    duplicate_product_names = products_df[products_df['product_name'].str.lower().duplicated(keep=False)]
    department_ids_of_duplicates = duplicate_product_names['department_id'].unique()
    print(department_ids_of_duplicates)


This shows the missing product names are not exclusively associated with department ID 21

In [ ]:
# What is this ailse and department?
# Filter rows with missing product names
missing_names = products_df[products_df['product_name'].isna()]

# Group by aisle and count the number of missing names
aisle_counts = missing_names.groupby('aisle_id')['product_name'].count()

# Group by department and count the number of missing names
dept_counts = missing_names.groupby('department_id')['product_name'].count()

print(aisle_counts)
print(dept_counts)

print(products_df['product_name'].isnull().sum())


After checking two different ways, this concludes it is not a specific aisle or department that has the missing products

In [ ]:
# Fill missing product names with 'Unknown'
products_df['product_name'].fillna('Unknown', inplace=True)
missing_products = products_df[products_df['product_name'].isnull()]
print(missing_products['product_name'])

I am a bit lost considering there are no missing product names in the product_df, but I did follow the coding procedure asked.

### `orders` data frame

In [ ]:
# Are there any missing values where it's not a customer's first order?
missing_not_first_order = orders_df[(orders_df['user_id'].isnull()) & (orders_df['order_id'] > 1)]
print(missing_not_first_order)
missing_not_first_order = orders_df[(orders_df['user_id'].isnull()) & (orders_df['order_id'] > 1)]
print(missing_not_first_order[['order_id', 'user_id']])
missing_values = orders_df.isnull().sum()
print(missing_values)

All three outputs I tried show there are no missing values where it's not a customers first order

### `order_products` data frame

In [ ]:
# What are the min and max values in this column?

min_value = order_products_df['add_to_cart_order'].min()
max_value = order_products_df['add_to_cart_order'].max()

print(f"Minimum value: {min_value}")
print(f"Maximum value: {max_value}")

Two of the columns are ID columns, meaning the min and max would not matter here, the reodered columns is binary so a min and max again does not make sense, this leads me to conclude that this is asking what the min and max values are of the add_to_cart_order column

In [ ]:
# Save all order IDs with at least one missing value in 'add_to_cart_order'
missing_order_ids = order_products_df[order_products_df['add_to_cart_order'].isnull()]['order_id'].unique()

print(missing_order_ids)


There are no missing orders IDs with atleast one missing values in the 'add_to_cart_order'

In [ ]:
# Do all orders with missing values have more than 64 products?

orders_with_missing_values = order_products_df[order_products_df['add_to_cart_order'].isnull()]

orders_with_more_than_64_products = orders_with_missing_values.groupby('order_id').size().reset_index(name='count')[lambda x: x['count'] > 64]

print(orders_with_more_than_64_products)

Since there are no missing values, there are none that have more than 64 products. But here is the code I would use in this instance if there were missing values

In [ ]:
# Replace missing values with ' ' and convert column to integer type
order_products_df['add_to_cart_order'].fillna('', inplace=True)
order_products_df['add_to_cart_order'] = order_products_df['add_to_cart_order'].astype(int)

Again, there were no missing values, but I did put the coding in that I would have done in the event that there was missing values.

# [A] Easy (must complete all to pass)

### [A1] Verify that the `'order_hour_of_day'` and `'order_dow'` values in the `orders` tables are sensible (i.e. `'order_hour_of_day'` ranges from 0 to 23 and `'order_dow'` ranges from 0 to 6)

In [ ]:
print("Minimum and maximum values of order_hour_of_day:")
print("Min:", orders_df['order_hour_of_day'].min())
print("Max:", orders_df['order_hour_of_day'].max())



In [ ]:
print("\nMinimum and maximum values of order_dow:")
print("Min:", orders_df['order_dow'].min())
print("Max:", orders_df['order_dow'].max())

### [A2] What time of day do people shop for groceries?

In [ ]:
import matplotlib.pyplot as plt

orders_df['order_hour_of_day'].hist(bins=24, edgecolor='black')
plt.xlabel('Hour of Day')
plt.ylabel('Number of Orders')
plt.title('Distribution of Orders by Hour of Day')
plt.show()

A bar graph seems the most convenient to show what hours of the day people buy their groceries. This displays that the peak hours that groceries are ordered are between 9am and 4pm.

### [A3] What day of the week do people shop for groceries?

In [ ]:
import matplotlib.pyplot as plt

days_of_week = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
orders_df['order_dow'].value_counts().plot(kind='bar', figsize=(10, 6))
plt.xlabel('Day of Week')
plt.ylabel('Number of Orders')
plt.title('Distribution of Orders by Day of Week')
plt.xticks(range(7), days_of_week)
plt.show()

The distribution of orders shows that Sunday and Monday are the busiest days for orders. 

### [A4] How long do people wait until placing another order?

In [ ]:

import matplotlib.pyplot as plt

orders_df['days_since_prior_order'].hist(bins=50, edgecolor='black')
plt.xlabel('Days Since Prior Order')
plt.ylabel('Number of Orders')
plt.title('Distribution of Days Since Prior Order')
plt.show()



It appears from first observations that the average time people wait to place an order is approximately once a week with an large outlier of people who wait once a month or more. Overall the grapg shows a mode of 30 days or more.

In [ ]:
print(orders_df['days_since_prior_order'].describe())
print()
# Calculate the median
median_days_since_prior_order = orders_df['days_since_prior_order'].median()
print(f"Median days since prior order: {median_days_since_prior_order:.2f} days")

Upon deeper inspection I found that the median amount of time is 7 days and the average amount of time between orders is approximately 11 day.

# [B] Medium (must complete all to pass)

### [B1] Is there a difference in `'order_hour_of_day'` distributions on Wednesdays and Saturdays? Plot the histograms for both days and describe the differences that you see.

In [ ]:
wednesdays = orders_df[orders_df['order_dow'] == 2]  # 2 corresponds to Wednesday
print(wednesdays)
plt.figure(figsize=(12, 6))


In [ ]:
saturdays = orders_df[orders_df['order_dow'] == 5]
print(saturdays)

In [ ]:
plt.figure(figsize=(12, 6))

plt.subplot(1, 2, 1)
plt.hist(wednesdays['order_hour_of_day'], bins=24, edgecolor='black')
plt.title('Order Hour of Day on Wednesdays')
plt.xlabel('Hour of Day')
plt.ylabel('Frequency')

plt.subplot(1, 2, 2)
plt.hist(saturdays['order_hour_of_day'], bins=24, edgecolor='black')
plt.title('Order Hour of Day on Saturdays')
plt.xlabel('Hour of Day')
plt.ylabel('Frequency')

plt.tight_layout()
plt.show()

In [ ]:

# Filter the data for Wednesdays and Saturdays
wednesdays = orders_df[orders_df['order_dow'] == 2]  # 2 corresponds to Wednesday
saturdays = orders_df[orders_df['order_dow'] == 5]  # 5 corresponds to Saturday

# Plot the histograms
plt.figure(figsize=(10, 6))

plt.hist(wednesdays['order_hour_of_day'], bins=24, alpha=0.5, label='Wednesdays', edgecolor='black')
plt.hist(saturdays['order_hour_of_day'], bins=24, alpha=0.5, label='Saturdays', edgecolor='black')

plt.title('Order Hour of Day on Wednesdays and Saturdays')
plt.xlabel('Hour of Day')
plt.ylabel('Frequency')
plt.legend()

plt.show()

The hour of the day that the orders are placed on Wednesday and Saturday vary just slightly but overall they follow the same trend per hour of the day.

### [B2] What's the distribution for the number of orders per customer?

In [ ]:
# Count the number of orders per customer
orders_per_customer = orders_df['user_id'].value_counts()

# Plot the distribution
plt.figure(figsize=(10, 6))
plt.hist(orders_per_customer, bins=50, edgecolor='black')
plt.title('Distribution of Orders per Customer')
plt.xlabel('Number of Orders')
plt.ylabel('Frequency')

plt.show()

In [ ]:
orders_per_customer = orders_df['user_id'].value_counts()
mean_orders_per_customer = orders_per_customer.mean()
print(f"The mean number of orders per customer is: {mean_orders_per_customer:.2f}")

The great majority of customers have under 10 times they have ordered groceries according to this data. The mean is 3 orders. 

### [B3] What are the top 20 popular products (display their id and name)?

In [ ]:
top_products = products_df['product_id'].value_counts().head(20)
print(top_products)

In [ ]:


top_products = products_df['product_id'].value_counts().head(20)
top_products_df = top_products.reset_index()
top_products_df.columns = ['product_id', 'order_count']

top_products_with_names = top_products_df.merge(products_df, on='product_id')
print(top_products_with_names[['product_id', 'product_name', 'order_count']].head(20))

The top 20 products and their names are displayed above. It appears the most popular item is Malbec Mendoza

# [C] Hard (must complete at least two to pass)

### [C1] How many items do people typically buy in one order? What does the distribution look like?

In [ ]:

items_per_order = order_products_df.groupby('order_id').size()
mean_items_per_order = items_per_order.mean()
print(f"The mean number of items per order is: {mean_items_per_order:.2f}")

In [ ]:
plt.hist(items_per_order, bins=50)
plt.xlabel('Number of items per order')
plt.ylabel('Frequency')
plt.title('Distribution of items per order')
plt.show()


The majority of distribution is under 20 items ordered at one time. There is no major outliers and the mean is 10.10 items.

### [C2] What are the top 20 items that are reordered most frequently (display their names and product IDs)?

In [ ]:
reordered_products_df = order_products_df[order_products_df['reordered'] == 1]
top_reordered_products = reordered_products_df.groupby('product_id')['product_id'].count().reset_index(name='reorder_count')
top_reordered_products = top_reordered_products.sort_values('reorder_count', ascending=False).head(20)



In [ ]:
# Merge with products_df to get product names
top_reordered_products_with_names = top_reordered_products.merge(products_df, on='product_id')



In [ ]:

print(top_reordered_products_with_names[['product_id', 'product_name', 'reorder_count']].head(20))

The most common reordered item was Bananas, product_id 24852. Bananas were reordered 55,763 times.

### [C3] For each product, what proportion of its orders are reorders?

In [ ]:
proportions = order_products_df.groupby('product_id')['reordered'].mean()
print(proportions)

In [ ]:
mean_reorder_percentage = proportions.mean()
print(f"Mean reorder percentage: {mean_reorder_percentage:.2f}%")

### [C4] For each customer, what proportion of their products ordered are reorders?

In [ ]:
# Merge the data on order_id
merged_data = pd.merge(order_products_df, orders_df, on='order_id')

# Group by user_id and calculate the total number of products ordered and reorders
user_data = merged_data.groupby('user_id').agg({
    'product_id': 'count',
    'reordered': 'sum'
}).reset_index()

# Rename the columns
user_data.columns = ['user_id', 'total_products', 'reorders']

# Calculate the proportion of reorders
user_data['proportion_reorders'] = user_data['reorders'] / user_data['total_products']

# Print the result
print(user_data)

In [ ]:
mean_proportion_reorders = user_data['proportion_reorders'].mean()
print(mean_proportion_reorders)

In [ ]:
plt.hist(user_data['proportion_reorders'], bins=20, alpha=0.5, label='Proportion of Reorders')

# Add a title and labels
plt.title('Distribution of Proportion of Reorders')
plt.xlabel('Proportion of Reorders')
plt.ylabel('Frequency')

# Show the mean proportion of reorders as a vertical line
plt.axvline(mean_proportion_reorders, color='r', linestyle='dashed', label='Mean')

# Show the legend
plt.legend()

# Show the plot
plt.show()

### [C5] What are the top 20 items that people put in their carts first? 

In [ ]:
# Get the first item added to each cart
first_items = order_products_df[order_products_df.groupby('order_id')['add_to_cart_order'].transform(min) == 1]



In [ ]:
# Get the top 20 products that people put in their carts first
top_20_first_items = first_items['product_id'].value_counts().head(20)
print(top_20_first_items)

In [ ]:
# Now you can merge the order_products DataFrame with the products DataFrame
merged_df = pd.merge(order_products_df, products_df, on='product_id')

# Get the first item added to each cart
first_items = merged_df[merged_df.groupby('order_id')['add_to_cart_order'].transform(min) == 1]

# Get the top 20 products that people put in their carts first
top_20_first_items = first_items['product_name'].value_counts().head(20)
print(top_20_first_items)

By merging order_products_df and products_df together, we are able to see what products people put in their cart first by its name and not just its product_id number. The list above shows the top 20 products put in the cart first. The leading product is bananas.

In [ ]:

# Create a horizontal bar chart of the top 20 first items
plt.barh(top_20_first_items.index, top_20_first_items.values)
plt.xlabel('Frequency')
plt.ylabel('Item')
plt.title('Top 20 Items Put in Cart First')
plt.show()

After cleaning up the data and exploring different ways to combine and look at the data,  I found that the busiest days for orders is  Sunday and Monday. The peak hours daily is between 10am and  4pm  and the  most popular item ordered and  reordered was bananas. Customers typically only order once or twice and most commonly  reorder every 30 days plus.